src/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMaterializer.cs

```diff
@@ -14,22 +14,41 @@ internal sealed class ReflectionEmitMaterializer : ClassMaterializer
         public override JsonClassInfo.ConstructorDelegate CreateConstructor(Type type)
         {
             Debug.Assert(type != null);
+            ConstructorInfo realMethod;
 
-            ConstructorInfo realMethod = type.GetConstructor(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance, binder: null, Type.EmptyTypes, modifiers: null);
-            if (realMethod == null)
+            if (type.IsValueType)
             {
-                return null;
+                realMethod = null;
+            }
+            else
+            {
+                realMethod = type.GetConstructor(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance, binder: null, Type.EmptyTypes, modifiers: null);
+                if (realMethod == null) return null;
             }
 
             var dynamicMethod = new DynamicMethod(
-                realMethod.Name,
-                type,
+                ".ctor",
+                typeof(object),
                 Type.EmptyTypes,
                 typeof(ReflectionEmitMaterializer).Module,
                 skipVisibility: true);
 
             ILGenerator generator = dynamicMethod.GetILGenerator();
-            generator.Emit(OpCodes.Newobj, realMethod);
+
+            if (type.IsValueType)
+            {
+                LocalBuilder local = generator.DeclareLocal(type);
+
+                generator.Emit(OpCodes.Ldloca_S, local);
+                generator.Emit(OpCodes.Initobj, type);
+                generator.Emit(OpCodes.Ldloc, local);
+                generator.Emit(OpCodes.Box, type);
```

Comment:
Does this have to be boxed?

Reply:
Yes, it boxes because `JsonClassInfo.ConstructorDelegate` has `object` as a return type. It's definitely possible to avoid boxing via preallocated `ByReference<T>` for each struct and storing it in a TLS or in a shared pool, but I intentionally haven't implemented that yet. The first step is basic support with tests, and optimization as the second after a discussion.